File: utils/Scene/sceneParser.cpp
    #pragma once

#include <cctype>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include "../../src/Vetor.h"
#include "../../src/Ponto.h"
#include "sceneSchema.hpp"
#include "jsonParser.cpp"


class SceneJsonLoader {
public:
    static SceneData loadFile(const std::string& filename) {
        Json root = parser.parseFile(filename);
        return build(root);
    }

    static SceneData loadString(const std::string& text) {
        Json root = parser.parseString(text);
        return build(root);
    }

private:
    static JsonParser parser;

    static SceneData makeEmptySceneData() {
        SceneData scene;

        scene.camera.lookfrom = Ponto(0, 0, 0);
        scene.camera.lookat   = Ponto(0, 0, -1);
        scene.camera.upVector = Vetor(0, 1, 0);
        scene.camera.image_width  = 0;
        scene.camera.image_height = 0;
        scene.camera.screen_distance = 1.0;

        scene.globalLight.pos = Ponto(0, 0, 0);
        scene.globalLight.color = ColorData(0, 0, 0);
        scene.globalLight.extraData.clear();

        scene.lightList.clear();
        scene.objects.clear();

        return scene;
    }

    static MaterialData makeDefaultMaterial(const std::string& name = "") {
        MaterialData m;
        m.name = name;
        m.color = ColorData(0.0, 0.0, 0.0);
        m.ks    = ColorData(0.0, 0.0, 0.0);
        m.ka    = ColorData(0.0, 0.0, 0.0);
        m.kr    = ColorData(0.0, 0.0, 0.0);
        m.kt    = ColorData(0.0, 0.0, 0.0);
        m.ns    = 0.0;
        m.ni    = 1.0;
        m.d     = 1.0;
        return m;
    }

    static bool isTripletArray(const Json& j) {
        return j.isArray() && j.size() == 3 &&
               j[0].isNumber() && j[1].isNumber() && j[2].isNumber();
    }

    static std::array<double, 3> readTriple(const Json& j, const std::string& what) {
        if (j.isArray()) {
            if (j.size() != 3) throw std::runtime_error("Expected 3 values for " + what);
            if (!j[0].isNumber() || !j[1].isNumber() || !j[2].isNumber()) throw std::runtime_error("Expected numeric 3-vector for " + what);
            return { j[0].asNumber(), j[1].asNumber(), j[2].asNumber() };
        }

        if (j.isObject()) {
            const auto& o = j.asObject();

            auto has = [&](const std::string& a, const std::string& b, const std::string& c) {
                return o.find(a) != o.end() && o.find(b) != o.end() && o.find(c) != o.end();
            };

            if (has("x", "y", "z")) return { o.at("x").asNumber(), o.at("y").asNumber(), o.at("z").asNumber() };
            if (has("r", "g", "b")) return { o.at("r").asNumber(), o.at("g").asNumber(), o.at("b").asNumber() };
        }

        throw std::runtime_error("Expected a 3-component vector for " + what);
    }

    static ColorData parseColor(const Json& j, const std::string& what = "color") {
        auto v = readTriple(j, what);
        return ColorData(v[0], v[1], v[2]);
    }

    static Vetor parseVector(const Json& j, const std::string& what = "vector") {
        auto v = readTriple(j, what);
        return Vetor(v[0], v[1], v[2]);
    }

    static Ponto parsePoint(const Json& j, const std::string& what = "point") {
        auto v = readTriple(j, what);
        return Ponto(v[0], v[1], v[2]);
    }

    static std::string requireString(const Json& obj, const std::string& key) {
        if (!obj.isObject() || !obj.contains(key) || !obj.at(key).isString()) throw std::runtime_error("Expected string field: " + key);
        return obj.at(key).asString();
    }

    static double requireNumber(const Json& obj, const std::string& key) {
        if (!obj.isObject() || !obj.contains(key) || !obj.at(key).isNumber()) throw std::runtime_error("Expected numeric field: " + key);
        return obj.at(key).asNumber();
    }

    static MaterialData parseMaterialObject(const Json& node, const std::string& fallbackName) {
        if (!node.isObject()) throw std::runtime_error("Material entry must be an object");

        MaterialData m = makeDefaultMaterial(fallbackName);

        if (node.contains("name") && node["name"].isString()) m.name = node["name"].asString();

        if (node.contains("color")) m.color = parseColor(node["color"], m.name + ".color");
        if (node.contains("ka"))    m.ka    = parseColor(node["ka"],    m.name + ".ka");
        if (node.contains("ks"))    m.ks    = parseColor(node["ks"],    m.name + ".ks");
        if (node.contains("kr"))    m.kr    = parseColor(node["kr"],    m.name + ".kr");
        if (node.contains("kt"))    m.kt    = parseColor(node["kt"],    m.name + ".kt");

        if (node.contains("ns") && node["ns"].isNumber()) m.ns = node["ns"].asNumber();
        if (node.contains("ni") && node["ni"].isNumber()) m.ni = node["ni"].asNumber();
        if (node.contains("d")  && node["d"].isNumber())  m.d  = node["d"].asNumber();

        return m;
    }

    static std::map<std::string, MaterialData> parseMaterials(const Json& materialsNode) {
        if (!materialsNode.isObject()) throw std::runtime_error("'definitions.materials' must be an object");

        std::map<std::string, MaterialData> table;

        for (const auto& [name, node] : materialsNode.asObject()) table[name] = parseMaterialObject(node, name);

        return table;
    }

    static MaterialData resolveMaterial(const Json& node, const std::map<std::string, MaterialData>& table) {
        if (node.isString()) {
            const std::string& ref = node.asString();
            auto it = table.find(ref);
            if (it == table.end()) throw std::runtime_error("Unknown material reference: " + ref);
            return it->second;
        }

        if (node.isObject()) return parseMaterialObject(node, node.contains("name") && node["name"].isString() ? node["name"].asString() : "");

        throw std::runtime_error("Invalid material format");
    }

    static CameraData parseCamera(const Json& node) {
        if (!node.isObject()) throw std::runtime_error("'camera' must be an object");

        CameraData cam;
        cam.lookfrom = Ponto(0, 0, 0);
        cam.lookat   = Ponto(0, 0, -1);
        cam.upVector = Vetor(0, 1, 0);
        cam.image_width = 800;
        cam.image_height = 800;
        cam.screen_distance = 1.0;

        if (node.contains("image_width"))  cam.image_width  = static_cast<int>(requireNumber(node, "image_width"));
        if (node.contains("image_height")) cam.image_height = static_cast<int>(requireNumber(node, "image_height"));

        if (node.contains("screen_distance")) cam.screen_distance = requireNumber(node, "screen_distance");

        if (node.contains("lookfrom")) cam.lookfrom = parsePoint(node["lookfrom"], "camera.lookfrom");
        if (node.contains("lookat"))   cam.lookat   = parsePoint(node["lookat"],   "camera.lookat");
        if (node.contains("vup"))      cam.upVector  = parseVector(node["vup"],     "camera.vup");

        return cam;
    }

    static LightData parseLight(const Json& node) {
        if (!node.isObject()) throw std::runtime_error("Each light must be an object");

        LightData light;
        light.pos = Ponto(0, 0, 0);
        light.color = ColorData(0, 0, 0);
        light.extraData.clear();

        if (node.contains("name") && node["name"].isString()) light.extraData["name"] = node["name"].asString();
        if (node.contains("position")) light.pos = parsePoint(node["position"], "light.position");
        if (node.contains("color")) light.color = parseColor(node["color"], "light.color");

        for (const auto& [key, val] : node.asObject()) {
            if (key == "name" || key == "position" || key == "color") continue;
            if (val.isString()) light.extraData[key] = val.asString();
            else if (val.isNumber()) light.extraData[key] = std::to_string(val.asNumber());
            else if (val.isBool()) light.extraData[key] = val.asBool() ? "true" : "false";
        }

        return light;
    }

    static TransformData parseTransform(const Json& node) {
        if (!node.isObject()) throw std::runtime_error("Each transform must be an object");

        TransformData t;
        t.tType = requireString(node, "type");

        for (const auto& [key, val] : node.asObject()) {
            if(key == "type") continue;
            t.data = parseVector(val, "transform." + key);
        }

        return t;
    }

    static std::vector<TransformData> parseTransforms(const Json& node) {
        if (!node.isArray()) throw std::runtime_error("'transform' must be an array");

        std::vector<TransformData> list;
        for (const auto& item : node.asArray())
            list.push_back(parseTransform(item));
        
        return list;
    }

    static ObjectData parseObject(const Json& node, const std::map<std::string, MaterialData>& materials) {
        if (!node.isObject()) throw std::runtime_error("Each object entry must be an object");

        ObjectData obj;
        obj.objType = "";
        obj.relativePos = Ponto(0, 0, 0);
        obj.material = makeDefaultMaterial();
        obj.numericData.clear();
        obj.vetorPointData.clear();
        obj.otherProperties.clear();
        obj.transforms.clear();


        auto isPositionKey = [](const std::string& key){
            static const std::vector<std::string> hints = {
                "center",
                "position",
                "point",
                "origin",
                "point_on_plane",
                "relativePos"
            };

            for (const auto& h : hints) {
                if (key == h)
                    return true;
            }
            return false;
        };
        
        if (node.contains("name") && node["name"].isString()) obj.otherProperties["name"] = node["name"].asString();
        if (node.contains("material")) obj.material = resolveMaterial(node["material"], materials);
        if (node.contains("transform")) obj.transforms = parseTransforms(node["transform"]);
        if (node.contains("relativePos")) obj.relativePos = parsePoint(node["relativePos"], "object.relativePos");
        if (node.contains("type") && node["type"].isString()) obj.objType = node["type"].asString();
        else throw std::runtime_error("Object missing required field: type");

        for (const auto& [key, val] : node.asObject()) {
            if (key == "type" || key == "material" || key == "transform" || key == "name" || key == "relativePos") continue;


            if (val.isNumber())      obj.numericData[key] = val.asNumber();
            else if (val.isString()) obj.otherProperties[key] = val.asString();
            else if (val.isBool())   obj.otherProperties[key] = val.asBool() ? "true" : "false";
            else if (isTripletArray(val)) {
                Vetor v = parseVector(val, "object." + key);
                obj.vetorPointData[key] = v;

                if(isPositionKey(key)) obj.relativePos = Ponto(v.getX(), v.getY(), v.getZ());
            }
            else  obj.otherProperties[key] = val.asString();
        }

        return obj;
    }

    static SceneData build(const Json& root) {
        if (!root.isObject()) {
            throw std::runtime_error("Root JSON value must be an object");
        }

        SceneData scene = makeEmptySceneData();
        std::map<std::string, MaterialData> materials;

        // globalLight
        if (root.contains("globalLight")) {
            scene.globalLight.color = parseColor(root["globalLight"], "globalLight");
        }

        // materials
        if (root.contains("materials")) {
            materials = parseMaterials(root["materials"]);
        }

        // camera
        if (root.contains("camera")) {
            scene.camera = parseCamera(root["camera"]);
        }

        // lights
        if (root.contains("lights")) {
            const Json& lights = root["lights"];
            if (!lights.isArray()) throw std::runtime_error("'lights' must be an array");
            
            for (const auto& item : lights.asArray()) {
                scene.lightList.push_back(parseLight(item));
            }
        }

        // objects
        if (root.contains("objects")) {
            const Json& objects = root["objects"];
            if (!objects.isArray()) throw std::runtime_error("'objects' must be an array");
            for (const auto& item : objects.asArray()) scene.objects.push_back(parseObject(item, materials));
        }

        return scene;
    }
};

// static member definition
inline JsonParser SceneJsonLoader::parser;

File: utils/Scene/sceneParser.py
from __future__ import annotations
import json
from pathlib import Path
from typing import Dict, List, Tuple

from src.Ponto import Ponto
from src.Vetor import Vetor
from utils.Scene.sceneSchema import (
    CameraData, ColorData, LightData, MaterialData,
    ObjectData, SceneData, TransformData,
)


class SceneJsonLoader:

    @staticmethod
    def load_file(filename: str) -> SceneData:
        path = Path(filename)
        if not path.exists():
            raise FileNotFoundError(f"Scene file not found: {filename}")
        with open(path, "r", encoding="utf-8") as f:
            root = json.load(f)
        return SceneJsonLoader._build(root)

    @staticmethod
    def load_string(text: str) -> SceneData:
        root = json.loads(text)
        return SceneJsonLoader._build(root)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _read_triple(value, what: str) -> Tuple[float, float, float]:
        """Aceita [a, b, c] ou {"x":..,"y":..,"z":..} ou {"r":..,"g":..,"b":..}."""
        if isinstance(value, list):
            if len(value) != 3:
                raise ValueError(f"Esperado 3 valores para {what}, obtido {len(value)}")
            return float(value[0]), float(value[1]), float(value[2])
        if isinstance(value, dict):
            if all(k in value for k in ("x", "y", "z")):
                return float(value["x"]), float(value["y"]), float(value["z"])
            if all(k in value for k in ("r", "g", "b")):
                return float(value["r"]), float(value["g"]), float(value["b"])
        raise ValueError(f"Esperado vetor de 3 componentes para {what}")

    @staticmethod
    def _parse_color(value, what="color") -> ColorData:
        r, g, b = SceneJsonLoader._read_triple(value, what)
        return ColorData(r, g, b)

    @staticmethod
    def _parse_vetor(value, what="vector") -> Vetor:
        x, y, z = SceneJsonLoader._read_triple(value, what)
        return Vetor(x, y, z)

    @staticmethod
    def _parse_ponto(value, what="point") -> Ponto:
        x, y, z = SceneJsonLoader._read_triple(value, what)
        return Ponto(x, y, z)

    @staticmethod
    def _is_triplet(value) -> bool:
        return (isinstance(value, list) and len(value) == 3
                and all(isinstance(v, (int, float)) for v in value))

    # ------------------------------------------------------------------
    # Seções do JSON
    # ------------------------------------------------------------------

    @staticmethod
    def _make_default_material(name: str = "") -> MaterialData:
        return MaterialData(
            name=name,
            color=ColorData(0, 0, 0),
            ks=ColorData(0, 0, 0),
            ka=ColorData(0, 0, 0),
            kr=ColorData(0, 0, 0),
            kt=ColorData(0, 0, 0),
            ns=0.0, ni=1.0, d=1.0,
        )

    @staticmethod
    def _parse_material_object(node: dict, fallback_name: str = "") -> MaterialData:
        m = SceneJsonLoader._make_default_material(fallback_name)
        if "name"  in node: m.name  = str(node["name"])
        if "color" in node: m.color = SceneJsonLoader._parse_color(node["color"], f"{m.name}.color")
        if "ka"    in node: m.ka    = SceneJsonLoader._parse_color(node["ka"],    f"{m.name}.ka")
        if "ks"    in node: m.ks    = SceneJsonLoader._parse_color(node["ks"],    f"{m.name}.ks")
        if "kr"    in node: m.kr    = SceneJsonLoader._parse_color(node["kr"],    f"{m.name}.kr")
        if "kt"    in node: m.kt    = SceneJsonLoader._parse_color(node["kt"],    f"{m.name}.kt")
        if "ns"    in node: m.ns    = float(node["ns"])
        if "ni"    in node: m.ni    = float(node["ni"])
        if "d"     in node: m.d     = float(node["d"])
        return m

    @staticmethod
    def _parse_materials(node: dict) -> Dict[str, MaterialData]:
        if not isinstance(node, dict):
            raise ValueError("'materials' deve ser um objeto JSON")
        return {
            name: SceneJsonLoader._parse_material_object(mat_node, name)
            for name, mat_node in node.items()
        }

    @staticmethod
    def _resolve_material(value, table: Dict[str, MaterialData]) -> MaterialData:
        if isinstance(value, str):
            if value not in table:
                raise ValueError(f"Material desconhecido: {value!r}")
            return table[value]
        if isinstance(value, dict):
            name = value.get("name", "")
            return SceneJsonLoader._parse_material_object(value, name)
        raise ValueError("Formato de material inválido (deve ser string ou objeto)")

    @staticmethod
    def _parse_camera(node: dict) -> CameraData:
        cam = CameraData()
        if "image_width"     in node: cam.image_width     = int(node["image_width"])
        if "image_height"    in node: cam.image_height    = int(node["image_height"])
        if "screen_distance" in node: cam.screen_distance = float(node["screen_distance"])
        if "lookfrom"        in node: cam.lookfrom = SceneJsonLoader._parse_ponto(node["lookfrom"], "camera.lookfrom")
        if "lookat"          in node: cam.lookat   = SceneJsonLoader._parse_ponto(node["lookat"],   "camera.lookat")
        # Suporta "upVector" (arquivos JSON) e "vup" (alias C++)
        for key in ("upVector", "vup", "up_vector"):
            if key in node:
                cam.up_vector = SceneJsonLoader._parse_vetor(node[key], f"camera.{key}")
                break
        return cam

    @staticmethod
    def _parse_light(node: dict) -> LightData:
        light = LightData()
        if "name"     in node and isinstance(node["name"], str):
            light.extra_data["name"] = node["name"]
        if "position" in node:
            light.pos   = SceneJsonLoader._parse_ponto(node["position"], "light.position")
        if "color"    in node:
            light.color = SceneJsonLoader._parse_color(node["color"],    "light.color")
        for key, val in node.items():
            if key in ("name", "position", "color"):
                continue
            if isinstance(val, str):
                light.extra_data[key] = val
            elif isinstance(val, (int, float)):
                light.extra_data[key] = str(val)
            elif isinstance(val, bool):
                light.extra_data[key] = "true" if val else "false"
        return light

    @staticmethod
    def _parse_transform(node: dict) -> TransformData:
        if "type" not in node:
            raise ValueError("Entrada de transform sem campo 'type'")
        t = TransformData(t_type=str(node["type"]))
        for key, val in node.items():
            if key == "type":
                continue
            t.data = SceneJsonLoader._parse_vetor(val, f"transform.{key}")
            break
        return t

    _POSITION_HINTS = frozenset({
        "center", "position", "point", "origin",
        "point_on_plane", "relativePos", "relative_pos",
    })

    @staticmethod
    def _parse_object(node: dict, materials: Dict[str, MaterialData]) -> ObjectData:
        if "type" not in node:
            raise ValueError("Objeto sem campo obrigatório: 'type'")

        obj = ObjectData(obj_type=str(node["type"]))

        if "name"        in node: obj.other_properties["name"] = str(node["name"])
        if "material"    in node: obj.material     = SceneJsonLoader._resolve_material(node["material"], materials)
        if "transform"   in node: obj.transforms   = [SceneJsonLoader._parse_transform(t) for t in node["transform"]]
        if "relativePos" in node: obj.relative_pos = SceneJsonLoader._parse_ponto(node["relativePos"], "object.relativePos")

        for key, val in node.items():
            if key in ("type", "material", "transform", "name", "relativePos"):
                continue
            if isinstance(val, bool):
                obj.other_properties[key] = "true" if val else "false"
            elif isinstance(val, (int, float)):
                obj.numeric_data[key] = float(val)
            elif isinstance(val, str):
                obj.other_properties[key] = val
            elif SceneJsonLoader._is_triplet(val):
                v = SceneJsonLoader._parse_vetor(val, f"object.{key}")
                obj.vetor_point_data[key] = v
                if key in SceneJsonLoader._POSITION_HINTS:
                    obj.relative_pos = Ponto(v.x, v.y, v.z)
            elif isinstance(val, list) or isinstance(val, dict):
                pass  # estruturas complexas ignoradas (ex: listas de faces inline)

        return obj

    # ------------------------------------------------------------------

    @staticmethod
    def _build(root: dict) -> SceneData:
        if not isinstance(root, dict):
            raise ValueError("Valor JSON raiz deve ser um objeto")

        scene = SceneData()
        materials: Dict[str, MaterialData] = {}

        if "globalLight" in root:
            scene.global_light.color = SceneJsonLoader._parse_color(root["globalLight"], "globalLight")

        if "materials" in root:
            materials = SceneJsonLoader._parse_materials(root["materials"])

        if "camera" in root:
            scene.camera = SceneJsonLoader._parse_camera(root["camera"])

        if "lights" in root:
            if not isinstance(root["lights"], list):
                raise ValueError("'lights' deve ser um array")
            scene.light_list = [SceneJsonLoader._parse_light(l) for l in root["lights"]]

        if "objects" in root:
            if not isinstance(root["objects"], list):
                raise ValueError("'objects' deve ser um array")
            scene.objects = [SceneJsonLoader._parse_object(o, materials) for o in root["objects"]]

        return scene
